### DISCUSSION TITLE
[10 Giugno 2022] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#55] [Cache (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/b0b7991a-af75-49e0-a691-7edc4a6cd989)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665874/AE.2021-22.-.Appello.1.-.2022-06-10.-.Esercizio.cache.-.Scheda.di.calcolo.pdf)

**Comment by daromacarlo** [comment]
Propongo soluzione per punto 5)

![tt](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/6cb1a404-9540-41e4-bee1-6615f57d6c06)

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/09fe2dc4-468d-4f65-9240-02a3c55987fe)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/63750b0f-748e-46a6-b127-2415bf4fd922)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1 + 1 + 27 + 2 * 4 * 8) * (3 * 4) \implies (29 + 64) * 12 \implies  93 * 12 \implies 1116\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 24 + 16 * 4 * 8) * 4 \implies (26 + 512) * 4\implies  537 * 4 \implies 2148\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 6 accessi in memoria, con tempo $\Large 400 * 6 = 2400\ \text{ns}$

- 2 accessi in L1, con tempo  $\Large 1 * 2 = 2\ \text{ns}$

- 4 accessi in L2, con tempo  $\Large 40 * 4 = 160\ \text{ns}$

$$
    \Large T_{tot} = 2400+ 2 + 160 = 2562\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{2562}{12} = 213.50\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{213.5 * 2}{4} = 106.75
$$

## 4) Word offset indirizzo 128 Cache L1

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (128 \gg 2)\ \\%\ 2^{1} \implies 32\ \\%\ 2 \implies 0
$$

## 5) TLB Hit & Page Miss

È impossibile che possa presentarsi il caso in cui si abbiamo sia TLB Hit che Page miss, in quanto la TLB non può contenere un'entrata il cui valid bit sia posto ad 1 e che sia relativa ad una pagina non caricata
